### ISSUE TITLE
`no-invalid-regexp` relies on parser to catch unsupported RegExp syntax features

### ISSUE THREAD

**Issue by kaicataldo** [eslint/eslint#13595]
Originally reported [here](https://github.com/babel/babel/issues/11915).

**What rule do you want to change?**

`no-invalid-regexp`

**Does this change cause the rule to produce more or fewer warnings?**

More (when using a third party parser that doesn't throw a syntax error for RegExp syntax features not supported in the `ecmaVersion` specified in the config)

**How will the change be implemented? (New option, new default behavior, etc.)?**

Taking a cursory look, we [hardcode `ecmaVersion: 2018`](https://github.com/eslint/eslint/blob/07db7b8080c2f68ee28e7d447db356c33e6fddce/lib/rules/no-invalid-regexp.js#L12) when we instantiate the RegExp validator. This is fine when using the default parser because it has knowledge of the supported `ecmaVersion` and throws an error at parse-time, but doesn't work when the parser doesn't have knowledge of the target supported syntax.

I think it should be the default behavior, but am open to discussing this. I would also argue that this should be treated as a semver-minor bug fix.

**Please provide some example code that this change will affect:**

With the following configuration:

```json
{
  "parser": "@babel/eslint-parser",
  "parserOptions": {
    "ecmaVersion": 2017,
  },
  "rules": {
    "no-invalid-regexp": "error",
  },
}
```

```js
new RegExp(/(?<!\?.+)(?<=\/)[\w-]+(?=[/\r\n?]|$)/)
```

**What does the rule currently do for this code?**

No warnings, because `@babel/eslint-parser` has no knowledge of the supported environment and doesn't throw a parser error.

**What will the rule do after it's changed?**

Validate the RegExp using `ecmaVersion: 2017`, which should warn about the use of negative lookbehind.

**Are you willing to submit a pull request to implement this change?**

Yes!

**Comment by mdjermanovic** [comment]
`no-invalid-regexp` uses [regexpp](https://github.com/eslint/eslint/blob/66442a9faf9872db4a40f56dde28c48f4d02fc7b/lib/rules/no-invalid-regexp.js#L11) parser, not the configured parser, so the parser from eslint config doesn't affect the behavior of this rule.

As for the original https://github.com/babel/babel/issues/11915 issue:

> ```js
> ```js
> new RegExp(/(?<!\?.+)(?<=\/)[\w-]+(?=[/\r\n?]|$)/)
> // SyntaxError: Invalid regular expression: invalid group specifier name
> ```
> 
> 
> With the default parser, setting parserOptions.ecmaVersion: 2017 and no-invalid-regexp: 2 correctly catches the error. With parser: 'babel-eslint' set, it seems to ignore the ecmaVersion and incorrectly asserts that the regex is valid.

`no-invalid-regexp` works with `string` arguments only. This error was reported by the parser: it's a regex _literal_  used in `RegExp` constructor call (which is usually redundant).

**Comment by mdjermanovic** [comment]
I guess the issue is that babel doesn't transform lookbehind assertions (because it probably isn't possible to do that), but there is no warning about that.

Maybe this rule could help:

https://eslint-plugin-es.mysticatea.dev/rules/no-regexp-lookbehind-assertions.html

**Comment by nzakas** [comment]
This has been open for three years without a clear path forward, so closing.
